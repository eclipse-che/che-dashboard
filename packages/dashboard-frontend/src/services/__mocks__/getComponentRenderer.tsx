/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
