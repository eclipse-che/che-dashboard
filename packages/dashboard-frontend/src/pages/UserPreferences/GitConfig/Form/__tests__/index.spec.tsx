/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { StateMock } from '@react-mock/state';
import userEvent from '@testing-library/user-event';
import React from 'react';

import getComponentRenderer, { screen, waitFor } from '@/services/__mocks__/getComponentRenderer';

import { GitConfigForm, Props, State } from '..';

jest.mock('@/pages/UserPreferences/GitConfig/Form/SectionUser');

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnSave = jest.fn();
const mockOnReload = jest.fn();

const defaultProps: Props = {
  isLoading: false,
  gitConfig: {
    user: {
      email: 'test@che',
      name: 'test',
    },
  },
  onSave: mockOnSave,
  onReload: mockOnReload,
};
const defaultState: State = {
  isValid: true,
  nextGitConfig: undefined,
};

describe('GitConfigForm', () => {
  test('snapshot', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  describe('buttons state', () => {
    test('initial', () => {
      renderComponent();

      expect(screen.getByRole('button', { name: 'Save' })).toBeDisabled();
      expect(screen.getByRole('button', { name: 'Reload' })).toBeEnabled();
    });

    test('while loading data', () => {
      renderComponent({ isLoading: true });

      expect(screen.getByRole('button', { name: 'Save' })).toBeDisabled();
      expect(screen.getByRole('button', { name: 'Reload' })).toBeDisabled();
    });

    test('with valid changes', () => {
      renderComponent(
        {},
        { isValid: true, nextGitConfig: { user: { name: 'new name', email: 'test@che' } } },
      );

      expect(screen.getByRole('button', { name: 'Save' })).toBeEnabled();
      expect(screen.getByRole('button', { name: 'Reload' })).toBeEnabled();
    });

    test('with invalid changes', () => {
      renderComponent(
        {},
        { isValid: false, nextGitConfig: { user: { name: '', email: 'test@che' } } },
      );

      expect(screen.getByRole('button', { name: 'Save' })).toBeDisabled();
      expect(screen.getByRole('button', { name: 'Reload' })).toBeEnabled();
    });

    test('handle valid changes', async () => {
      renderComponent();

      // state with no changes
      expect(screen.getByRole('button', { name: 'Save' })).toBeDisabled();
      expect(screen.getByRole('button', { name: 'Reload' })).toBeEnabled();

      const changeEmailButton = screen.getByRole('button', { name: 'Change Email Valid' });
      await userEvent.click(changeEmailButton);

      // state with valid changes
      expect(screen.getByRole('button', { name: 'Save' })).toBeEnabled();
      expect(screen.getByRole('button', { name: 'Reload' })).toBeEnabled();
    });
  });

  test('handle the Reload button click', async () => {
    renderComponent(
      {},
      { isValid: true, nextGitConfig: { user: { name: 'new name', email: 'test@che' } } },
    );

    // expect the name to be changed
    expect(screen.getByTestId('user-name')).toHaveTextContent('new name');

    const reloadButton = screen.getByRole('button', { name: 'Reload' });
    await userEvent.click(reloadButton);

    expect(mockOnReload).toHaveBeenCalled();

    // expect the name to be reverted
    await waitFor(() => expect(screen.getByTestId('user-name')).toHaveTextContent('test'));
  });

  test('handle the Save button click', () => {
    renderComponent(
      {},
      { isValid: true, nextGitConfig: { user: { name: 'new name', email: 'test@che' } } },
    );

    screen.getByRole('button', { name: 'Save' }).click();

    expect(mockOnSave).toHaveBeenCalledWith({ user: { name: 'new name', email: 'test@che' } });
  });
});

function getComponent(props: Partial<Props> = {}, state: Partial<State> = {}): React.ReactElement {
  const localState = { ...defaultState, ...state };
  return (
    <StateMock state={localState}>
      <GitConfigForm {...defaultProps} {...props} />
    </StateMock>
  );
}
