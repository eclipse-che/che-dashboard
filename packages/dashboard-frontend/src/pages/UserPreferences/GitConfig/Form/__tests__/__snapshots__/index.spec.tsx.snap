// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitConfigForm snapshot 1`] = `
<section
  className="pf-c-page__main-section pf-m-light"
>
  <form
    className="pf-c-form pf-m-horizontal"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div>
      <div
        data-testid="user-email"
      >
        test@che
      </div>
      <div
        data-testid="user-name"
      >
        test
      </div>
      <button
        onClick={[Function]}
      >
        Change Email Valid
      </button>
    </div>
    <div
      className="pf-l-flex"
    >
      <button
        aria-disabled={true}
        aria-label={null}
        className="pf-c-button pf-m-primary pf-m-disabled"
        data-ouia-component-id="OUIA-Generated-Button-primary-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        data-testid="button-save"
        disabled={true}
        onClick={[Function]}
        role={null}
        tabIndex={null}
        type="button"
      >
        Save
      </button>
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-secondary"
        data-ouia-component-id="OUIA-Generated-Button-secondary-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        data-testid="button-reload"
        disabled={false}
        onClick={[Function]}
        role={null}
        type="reset"
      >
        Reload
      </button>
    </div>
  </form>
</section>
`;
